Mathlib/CategoryTheory/Sites/Coherent/RegularSheaves.lean

```diff
@@ -54,9 +56,147 @@ def EqualizerCondition (P : Cᵒᵖ ⥤ Type*) : Prop :=
     (MapToEqualizer P π (pullback.fst (f := π) (g := π)) (pullback.snd (f := π) (g := π))
     pullback.condition)
 
+lemma equalizerCondition_iff_isIso_lift_w (P : Cᵒᵖ ⥤ Type*) {X B : C} (π : X ⟶ B)
+    [HasPullback π π] : P.map π.op ≫ P.map (pullback.fst (f := π) (g := π)).op =
+    P.map π.op ≫ P.map (pullback.snd).op := by
+  simp only [← Functor.map_comp, ← op_comp, pullback.condition]
+
+lemma mapToEqualizer_eq_comp (P : Cᵒᵖ ⥤ Type*) {X B : C} (π : X ⟶ B)
+    [HasPullback π π] : MapToEqualizer P π pullback.fst pullback.snd pullback.condition =
+    equalizer.lift (P.map π.op) (equalizerCondition_iff_isIso_lift_w P π) ≫
+    (Types.equalizerIso _ _).hom := by
+  rw [← Iso.comp_inv_eq (α := Types.equalizerIso _ _)]
+  apply equalizer.hom_ext
+  aesop
+
+theorem equalizerCondition_iff_isIso_lift (P : Cᵒᵖ ⥤ Type*) : EqualizerCondition P ↔
```

Comment:
In order to have an API which parallels what we have for `PreservesLimit`, could you change the definition of `EqualizerCondition` so that it becomes:
```lean
def EqualizerCondition (P : Cᵒᵖ ⥤ Type*) : Prop :=
  ∀ ⦃X B : C⦄ (π : X ⟶ B) [EffectiveEpi π] (c : PullbackCone π π) (_ : IsLimit c),
    Nonempty (IsLimit (P.mapCone (PushoutCocone.mk _ _ c.condition).op))
```
(Actually, you may also replace `Type*` by any category, which should allow handling (pre)sheaves in any category directly.)

(It would makes sense to give a name to this definition `c => PushoutCocone.mk _ _ c.condition`, and maybe develop the relation with this construction with `EffectiveEpi`: show that if a pullback cocone like this is limit, then corresponding square is both cartesian and cocartesian.)

Then, I would think that when translated in the language of general pushout/pullbackback co/cones, the code would eventually look nicer.

Reply:
More precisely, modulo a small refactor of `isLimitMapConePullbackConeEquiv` so that it allows any `PullbackCone` rather than those constructed by `PullbackCone.mk`, and a general compatibility of pullback squares with isos, the proof of `equalizerCondition_of_natIso` would become much easier:
```lean
section

variable {D : Type*} [Category D] (G : C ⥤ D) {X Y Z : C} {f : X ⟶ Z} {g : Y ⟶ Z}
  (c : PullbackCone f g)

def isLimitMapConePullbackConeEquiv :
    IsLimit (G.mapCone c) ≃
      IsLimit (PullbackCone.mk (G.map c.fst) (G.map c.snd)
      (by simpa using G.congr_map c.condition) : PullbackCone (G.map f) (G.map g)) := sorry
end

def PullbackCone.isLimitOfIsos {X₁ X₂ S : C} {f₁ : X₁ ⟶ S} {f₂ : X₂ ⟶ S}
    (c : PullbackCone f₁ f₂) (hc : IsLimit c) {X₁' X₂' S' : C}
    {f₁' : X₁' ⟶ S'} {f₂' : X₂' ⟶ S'} (c' : PullbackCone f₁' f₂')
    (e₀ : S ≅ S') (e₁ : X₁ ≅ X₁') (e₂ : X₂ ≅ X₂') (e : c.pt ≅ c'.pt)
    (comm₁ : e₁.hom ≫ f₁' = f₁ ≫ e₀.hom := by aesop_cat)
    (comm₂ : e₂.hom ≫ f₂' = f₂ ≫ e₀.hom := by aesop_cat)
    (comm₃ : e.hom ≫ c'.fst = c.fst ≫ e₁.hom := by aesop_cat)
    (comm₄ : e.hom ≫ c'.snd = c.snd ≫ e₂.hom := by aesop_cat) : IsLimit c' := sorry

theorem equalizerCondition_of_natIso {P P' : Cᵒᵖ ⥤ A} (i : P ≅ P')
    (hP : EqualizerCondition P) : EqualizerCondition P' := fun {X B} π _ c hc =>
  ⟨(isLimitMapConePullbackConeEquiv P' _).2
    (PullbackCone.isLimitOfIsos _ ((isLimitMapConePullbackConeEquiv P _).1 (hP π c hc).some) _
    (i.app _) (i.app _) (i.app _) (i.app _))⟩
```

Reply:
Thanks for the suggestions! I didn't do the `mapCone` version, but a slightly different one which worked quite well. It also uses general pullback cones.